markup.go

```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
 	Property("className", strings.Join(classes, " ")).Apply(h)
 }
 
-// List represents a list of Markup, Component, or HTML which is individually
+// MarkupList represents a list of Applyer which is individually
 // applied to an HTML element or text node.
-type List []MarkupOrComponentOrHTML
+type MarkupList []Applyer
 
-// Apply implements the Markup interface.
-func (l List) Apply(h *HTML) {
-	for _, m := range l {
-		apply(m, h)
+// Apply implements the Applyer interface.
+func (m MarkupList) Apply(h *HTML) {
+	for _, a := range m {
+		if a == nil {
+			continue
+		}
+		a.Apply(h)
 	}
 }
 
+// Markup wraps a list of Applyer which is individually
+// applied to an HTML element or text node.
+func Markup(m ...Applyer) MarkupList {
+	return MarkupList(m)
+}
+
+// List wraps a list of Markup, Component, or HTML which is individually
+// applied to an HTML element or text node.
+func List(m ...ComponentOrHTML) ChildList {
```

Comment:
It's possible for a user to shoot themselves in the foot here due to the empty interface types, and get a panic at render time, like so:

```go
var blah []ComponentOrHTML
vecty.List(blah)
```
When they should instead do:
```go
var blah []ComponentOrHTML
vecty.List(blah...)
```

Reply:
We could alternatively just retain:
```go
type List []ComponentOrHTML
```
instead of `ChildList`, but it felt odd that `Markup` was a function, whereas `List` was a type, so they would require different syntax to use.

Reply:
> We could alternatively just retain: ...

The reason `Markup` is a function is because the most common use case for `Markup` is something like this:

```
return elem.Div(
	vecty.Markup(
		vecty.Style("float", "right"),
	),
	...
}
```

So that users don't have to write:

```
return elem.Div(
	[]vecty.Markup{
		vecty.Style("float", "right"),
	},
	...
}
```

Which wouldn't really match the style of other things inside a component's `Render` method.

---

In contrast to the above, `List` is _not_ commonly used inline inside a `elem.Foobar()` function call. For example, I do not really expect users to write:

```
return elem.Div(
	vecty.List(
		vecty.Text("Hello"),
		vecty.Text("World"),
	),
}
```

Because they could instead just write that as:

```
return elem.Div(
	vecty.Text("Hello"),
	vecty.Text("World"),
}
```

What I would expect users to do is use `vecty.List` in place of a regular `[]ComponentOrHTML`, for example in code like https://github.com/pdf/vecty/blob/e71f4c2cda8a9fcd493e78d1034bb9befed0ab2e/example/todomvc/components/pageview.go#L179-L185, because writing `ComponentOrHTML` is tedious. IMHO for this use case, `vecty.List()` being a function seems more ambiguous as to what it really does.

---

Because of the above, I would say that we should just retain:

```
type List []ComponentOrHTML
```

Also, with the above, they can't shoot themselves in the foot with at least that specific example:

```
var blah []ComponentOrHTML
vecty.List(blah)
```

because that is now a type conversion. Of course, there are other various ways they could still shoot themselves in the foot, though.

Reply:
The same applies foot-shooting caveat applies to the Markup function though:

```go
var markup []Applyer
vecty.Markup(markup) // wrong
```

The awkwardness though was that if users have a children var of `[]ComponentOrHTML`, and a markup var of `[]Applyer`, the syntax to use them is different if we retain the old `List` type:

```go
var children []ComponentOrHTML
var markup []Applyer

vecty.Markup(markup...)
vecty.List(children)
```

I... think I'm okay with changing `List` though...

Reply:
Let's ignore the foot-shooting portion: I agree it occurs in both cases.

---

I'm aware of the awkwardness you describe, but you're assuming that users will be converting their own `children` list via `vecty.List(children)`, why? i.e. with this change, you are saying that code previously written as:

```
var children vecty.List
children = append(children, foobar)
return Div(
    children,
)
```

should instead become:

```
var children []ComponentOrHTML
children = append(children, foobar)
return Div(
    vecty.List(children),
)
```

Sure, the above could just be `var children vecty.ChildList` -- but that is kind-of my point: I don't see in which scenario people would use this `vecty.List` helper function?

Reply:
ack, I'll change this in the next pass.